bestbook/api/books.py

```diff
@@ -64,105 +63,113 @@ class BookGraph(core.Base):
     created = Column(DateTime(timezone=False), default=datetime.utcnow,
                      nullable=False)
     modified = Column(DateTime(timezone=False), default=None)
 
+    # topic = relationship("Topic")
 
-    @staticmethod
-    def clean_olid(olid):
-        """
-        Extract just the olid from some olid-containing string
-
-        >>> Book.clean_olid(None)
-        ''
-        >>> Book.clean_olid('')
-        ''
-        """
-        if not olid:
-            return ''
-        olid = olid.upper()
-        if olid.startswith('OL') and olid.endswith(('M', 'W')):
-            return olid
-        return re.findall(r'OL[0-9]+[MW]', olid)[0]
-
-    @staticmethod
-    def get_many(olids):
-        if not olids:
-            return {}
-        q = 'key:(%s)' % ' OR '.join(
-            '/%s/%s' % (
-                'works' if olid.endswith('W') else 'editions',
-                olid
-            ) for olid in olids
-        )
-        url = '%s/search.json?q=%s' % (OL_API, q)
-        r = requests.get(url)
+    # TODO: Create relationship to Bookgraph entry (or entries)
+    # This will allow us to remove submitter, topic_id, and winner_work_olid
+    nodes = relationship("BookGraph")
 
-        books = {}
-        try:
-            r.raise_for_status()
-        except requests.exceptions.HTTPError:
-            return books
+    # TODO: Replace the following with data from book graph:
+    winner = None
+    # winner = relationship("Book", backref="recommendations", foreign_keys=[book_id])
+    # candidates = relationship("BookGraph", secondary='recommendations_to_books')
 
-        for doc in r.json().get('docs', {}):
-            books[doc.get('key', '').split('/')[-1]] = doc
-        return books
+    @classmethod
+    def topics(cls):
+        topic_ids = [t.topic_id for t in cls.query.distinct(cls.topic_id)]
+        return Topic.query.filter(Topic.id.in_(topic_ids)).all()
 
     @classmethod
-    def get_work_and_edition(cls, olid):
-        work_id = None
-        edition_id = None
-        if olid.lower().endswith('m'):
-            edition_id = olid
-            # Fetch the edition's corresponding work_id from Open Library
-            r = requests.get('%s/books/%s' % (OL_API,  olid)).json()
-            work_id = r['works'][0]['key'].split('/')[-1]
-        else:
-            work_id = olid
-        return (
-            work_id and cls.clean_olid(work_id),
-            edition_id and cls.clean_olid(edition_id)
-        )
+    def paginate(cls, page, limit=10, **kwargs):
+        olids = []
+        revs = cls.query.filter_by(**kwargs).limit(limit).offset(page * limit)
+        for r in revs:
+            for n in r.nodes:
+                olids.append(n.contender_work_olid)
+            olids.append(n.winner_work_olid)
+            r.winner = n.winner_work_olid
+        revs.works = get_many(olids)
+        return revs
+
+    def get_topic(self):
+        return self.nodes[0].topic
```

Comment:
Are we sure that `self.nodes` will _never_ be empty?

Reply:
Why not:
```
    @property
    def topic(self):
        return self.nodes[0].topic
```

Reply:
> Are we sure that self.nodes will never be empty?

Ideally yes, but I didn't take SQL errors into account...

All related book graph and review table entries are created at the same time, but not in a single transaction (the same is true for deletion).  I think that it would be _fairly_ safe to assume that `self.nodes` will not be empty if the inserts and deletes occur within a single transaction.

Reply:
Opted to use `@hybrid_property`, as properties couldn't be accessed from templates while using `@property`.